### ISSUE TITLE
can't buid with gcc-5.3.0 and waf 1.9.7

### ISSUE THREAD

**Issue by TkrUdagawa** [jubatus/jubatus_core#349]
I tried building jubatus_core 1.0.1-p1 with gcc-5.3.0 but failed.
The error log was below:

```
[183/213] Compiling jubatus/core/nearest_neighbor/lsh_function.cpp
{standard input}: Assembler messages:
{standard input}:22469: Error: suffix or operands invalid for `vpsrld'
{standard input}:22485: Error: suffix or operands invalid for `vpsrld'
{standard input}:22580: Error: suffix or operands invalid for `vpsrld'
{standard input}:22594: Error: suffix or operands invalid for `vpsubd'
{standard input}:22743: Error: suffix or operands invalid for `vpaddd'
{standard input}:22744: Error: suffix or operands invalid for `vpand'
{standard input}:22758: Error: suffix or operands invalid for `vpand'
{standard input}:22765: Error: suffix or operands invalid for `vpand'
{standard input}:22771: Error: suffix or operands invalid for `vpsubd'
{standard input}:22798: Error: suffix or operands invalid for `vpslld'
{standard input}:22826: Error: suffix or operands invalid for `vpandn'
{standard input}:22832: Error: suffix or operands invalid for `vpcmpeqd'
{standard input}:22874: Error: suffix or operands invalid for `vpslld'
```

Next I checkout the waf of 1.0.0-p1 (waf version 1.6.4) and built again.
Then succeeded.

Other machine information is below:
```
OS: CentOS release 6.7 (Final)
CPU: AMD Opteron(TM) Processor 6276
```

I'm not still sure what conditions cause this error.

**Comment by kazuki** [comment]
please tell me output of "./waf configure"

**Comment by kazuki** [comment]
wafのバージョンによってビルドに成功するのは，
waf configureで両方共 fmv = yes となっているにも関わらず，
waf-1.9.7でしか -DJUBATUS_USE_FMV=1 が利用されていなからのようです．
(おそらく wscript or waf のバグ？)

waf-1.6.4
```
[152/193] cxx: jubatus/core/nearest_neighbor/lsh_function.cpp -> build/jubatus/core/nearest_neighbor/lsh_function.cpp.1.o
12:14:04 runner ['/opt/lib/bin/g++', '-O2', '-Wall', '-g', '-pipe', '-fno-omit-frame-pointer', '-pthread', '-fPIC', '-I/home/jubatus/jubatus_core/build', '-I/home/jubatus/jubatus_core', '-DJUBATUS_CORE_VERSION="1.0.0"', '-DJUBATUS_CORE_APPNAME="jubatus_core"', '-DNDEBUG=1', '-DJUBATUS_DISABLE_ASSERTIONS=1', '-DBUILD_\
DIR="/home/jubatus/jubatus_core/build"', '-DJUBATUS_USE_EIGEN=1', '-DHAVE_ONIGURUMA=1', '../jubatus/core/nearest_neighbor/lsh_function.cpp', '-c', '-o', 'jubatus/core/nearest_neighbor/lsh_function.cpp.1.o']
```

waf-1.9.7
```
12:15:43 runner ['/opt/lib/bin/g++', '-O2', '-Wall', '-g', '-pipe', '-fno-omit-frame-pointer', '-pthread', '-fPIC', '-I.', '-I..', '-DJUBATUS_CORE_VERSION="1.0.1"', '-DJUBATUS_CORE_APPNAME="jubatus_core"', '-DJUBATUS_PLUGIN_DIR="/usr/local/lib64/jubatus/plugin"', '-DNDEBUG=1', '-DJUBATUS_DISABLE_ASSERTIONS=1', '-DBUI\
LD_DIR="/home/jubatus/jubatus_core/build"', '-DJUBATUS_USE_EIGEN=1', '-DJUBATUS_USE_FMV=1', '-DHAVE_ONIGURUMA=1', '-DHAVE_DL=1', '-DLIBSPLITTER_SAMPLE="./libsplitter_sample.so"', '-DLIBSPLITTER_NULL="./libsplitter_null.so"', '-DLIBFILTER_SAMPLE="./libfilter_sample.so"', '-DLIBNUM_FEATURE_SAMPLE="./libnum_feature_samp\
le.so"', '-DLIBNUM_FILTER_SAMPLE="./libnum_filter_sample.so"', '-DLIBBINARY_FEATURE_SAMPLE="./libbinary_feature_sample.so"', '../jubatus/core/nearest_neighbor/lsh_function.cpp', '-c', '-o/home/jubatus/jubatus_core/build/jubatus/core/nearest_neighbor/lsh_function.cpp.2.o']
    {task 139668668356320: cxx lsh_function.cpp -> lsh_function.cpp.2.o}
['/opt/lib/bin/g++', '-O2', '-Wall', '-g', '-pipe', '-fno-omit-frame-pointer', '-pthread', '-fPIC', '-I.', '-I..', '-DJUBATUS_CORE_VERSION="1.0.1"', '-DJUBATUS_CORE_APPNAME="jubatus_core"', '-DJUBATUS_PLUGIN_DIR="/usr/local/lib64/jubatus/plugin"', '-DNDEBUG=1', '-DJUBATUS_DISABLE_ASSERTIONS=1', '-DBUILD_DIR="/home/ju\
batus/jubatus_core/build"', '-DJUBATUS_USE_EIGEN=1', '-DJUBATUS_USE_FMV=1', '-DHAVE_ONIGURUMA=1', '-DHAVE_DL=1', '-DLIBSPLITTER_SAMPLE="./libsplitter_sample.so"', '-DLIBSPLITTER_NULL="./libsplitter_null.so"', '-DLIBFILTER_SAMPLE="./libfilter_sample.so"', '-DLIBNUM_FEATURE_SAMPLE="./libnum_feature_sample.so"', '-DLIBN\
UM_FILTER_SAMPLE="./libnum_filter_sample.so"', '-DLIBBINARY_FEATURE_SAMPLE="./libbinary_feature_sample.so"', '../jubatus/core/nearest_neighbor/lsh_function.cpp', '-c', '-o/home/jubatus/jubatus_core/build/jubatus/core/nearest_neighbor/lsh_function.cpp.2.o']
```

エラーの内容はAVX2の命令なのですが，FMVはビルド環境がその命令セットを備えなくても良いはずですし，waf configureでのFMVのチェックはパスしているので，どうしてここでエラーになるのかは不明です．(gccのバグ？)

対応する必要があるのであれば，
Jubatusが対応する環境以外で古いgccを使う場合向けに，configureに--disable-fmvというオプションをwscriptに追加するなどでしょうか？

**Comment by kmaehashi** [comment]
I tested this in CentOS release 6.7 (Final) Docker container on Ubuntu 14.04.5 host (Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz ) with manually-installed gcc-5.3.0 (with following libs) and Jubatus Core 1.0.3, and it succeeds...

```
gcc-5.3.0
gmp-5.0.5
mpc-1.0
mpfr-3.1.1
ppl-0.11.2
```

@TkrUdagawa Are you on VM, Docker or bare host?

**Comment by kazuki** [comment]
@kmaehashi Core i7-4790 implements AVX2 instructions.
https://ark.intel.com/ja/products/80806/Intel-Core-i7-4790-Processor-8M-Cache-up-to-4_00-GHz

This problem occurs only on a CPU that does not implement the AVX 2 instruction.

**Comment by kmaehashi** [comment]
Hmm, maybe my waf configure cache was strange.
Running `./waf clean; ./waf configure; ./waf build --checkall` reproduced the error even on Core i7-4790 box.

```
[182/213] Compiling jubatus/core/nearest_neighbor/lsh_function.cpp
{standard input}: Assembler messages:
{standard input}:22469: Error: suffix or operands invalid for `vpsrld'
{standard input}:22485: Error: suffix or operands invalid for `vpsrld'
{standard input}:22580: Error: suffix or operands invalid for `vpsrld'
{standard input}:22594: Error: suffix or operands invalid for `vpsubd'
{standard input}:22743: Error: suffix or operands invalid for `vpaddd'
{standard input}:22744: Error: suffix or operands invalid for `vpand'
{standard input}:22758: Error: suffix or operands invalid for `vpand'
{standard input}:22765: Error: suffix or operands invalid for `vpand'
{standard input}:22771: Error: suffix or operands invalid for `vpsubd'
{standard input}:22798: Error: suffix or operands invalid for `vpslld'
{standard input}:22826: Error: suffix or operands invalid for `vpandn'
{standard input}:22832: Error: suffix or operands invalid for `vpcmpeqd'
{standard input}:22874: Error: suffix or operands invalid for `vpslld'
```

**Comment by kmaehashi** [comment]
Another issue I faced is that `ld.gold` is unavailable by default when manually installing gcc-5.3.0, so `./waf configure` fails.

**Comment by kmaehashi** [comment]
After investigation and some googling, I finally believe that this issue occurs because binutils (GNU Assembler, or ``as``) bundled with RHEL 6 is too old, which is, ``2.20.51.0.2-5.47.el6_9.1``.


Facts:

* Although the minimum required binutils version is not clearly written on the GCC changelog, it seems that GCC 5.x series expects binutils 2.25 or similar. https://gcc.gnu.org/gcc-5/changes.html
* GCC 5.3.1 included in ``devtoolset-4`` of Software Collections (SCL) comes with binutils ``2.25.1-8.el6``.  Under ``devtoolset-4``, Jubatus builds successfully without error.
* I found some people saying binutils from RHEL 6 is too old and does not fully support Haswell/AVX2 (I'm not sure about this).  https://github.com/hpcugent/easybuild-easyconfigs/issues/1605#issuecomment-103042773 http://www.pgroup.com/userforum/viewtopic.php?t=4502&view=next&sid=57ed7fc850f23a24014889d41e1058f7

So my suggestion is to update your binutils if you use recent versions of GCC.

I think it's a hard way (and thus should be generally avoided) to manually build GCC, at least for the rest of us...

**Comment by kazuki** [comment]
Thanks!

binutils/gas supports AVX2 in this commit https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commit;h=6c30d220f19d7f4fe197fa5f62a472b2e5e84d7c

binutils-2.22 and later, supports AVX2.
Maybe, I think most of users using the new gcc are using the new binutils... 
We need to add binutils version check script in wscript ?

```
$ git tag -l --contain 6c30d220f19d7f4fe197fa5f62a472b2e5e84d7c 'binutils*'
binutils-2_22
binutils-2_22-branchpoint
binutils-2_23
binutils-2_23-branchpoint
binutils-2_23_1
binutils-2_23_2
binutils-2_24
binutils-2_24-branchpoint
binutils-2_25
binutils-2_25_1
binutils-2_26
binutils-2_26_1
binutils-2_27
binutils-2_28
```

**Comment by kmaehashi** [comment]
Thanks for checking binutils version details!

>  We need to add binutils version check script in wscript ?

Currently ``func_multiver_test_code`` is using ``_mm256_xor_si256`` (VPXOR) for AVX2 test.
Adding ``_mm256_srli_epi`` (VPSRLD) test may help in this case.

Red Hat (or other distributors) may be backporting some code from future, so we cannot reliably tell what is supported and what is not from the version number.

**Comment by kmaehashi** [comment]
For later reference:

* ``VPXOR`` is available in AVX (and AVX2): https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=VPXOR&techs=AVX
* ``VPSRLD`` is not available in AVX (first appear in AVX2): https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=VPSRLD&techs=AVX

**Comment by kmaehashi** [comment]
Fixed via #373
